feat(analysis/transcendental): e is transcendental

Ruben-VandeVelde opened the pull request:
---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message 
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

#3884, second attempt.

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

jjaassoonn committed:
transcendental_things

Ruben-VandeVelde committed:
fixup! transcendental_things

Ruben-VandeVelde committed:
fixup! transcendental_things

Ruben-VandeVelde committed:
fixup! transcendental_things

Ruben-VandeVelde committed:
fixup! transcendental_things

Ruben-VandeVelde committed:
fixup! transcendental_things

Ruben-VandeVelde committed:
fixup! transcendental_things

Ruben-VandeVelde committed:
fixup! transcendental_things

Ruben-VandeVelde committed:
fixup! transcendental_things

Ruben-VandeVelde committed:
fixup! transcendental_things

Ruben-VandeVelde committed:
fixup! transcendental_things

Ruben-VandeVelde committed:
fixup! transcendental_things

Ruben-VandeVelde committed:
fixup! transcendental_things

Ruben-VandeVelde committed:
fixup! transcendental_things

Ruben-VandeVelde committed:
fixup! transcendental_things

Ruben-VandeVelde committed:
fixup! transcendental_things

Ruben-VandeVelde committed:
fixup! transcendental_things

Ruben-VandeVelde committed:
fixup! transcendental_things

Ruben-VandeVelde committed:
fixup! transcendental_things

Ruben-VandeVelde committed:
Gather deriv lemmas

vihdzp commented on src/analysis/transcendental/basic.lean:
```diff
@@ -0,0 +1,6 @@
+
+theorem transcendental_iff_transcendental_over_ℚ (x : ℝ) :
+  transcendental ℤ x ↔ transcendental ℚ x :=
+iff.not $ is_fraction_ring.is_algebraic_iff ℤ ℚ ℝ
```
```suggestion
(is_fraction_ring.is_algebraic_iff ℤ ℚ ℝ).not
```

vihdzp commented on src/analysis/transcendental/deriv.lean:
```diff
@@ -0,0 +1,247 @@
+/-- # Definition and some theorems about differentiating multiple times
+-/
+
+/-Definition
```
Note, there's no need to specify whether a definition/theorem is that. I don't think this is done anywhere else in mathlib.

vihdzp commented on src/analysis/transcendental/deriv.lean:
```diff
@@ -0,0 +1,247 @@
+/-Definition
+For any integer polynomial $f$ and $n\in\mathbb N$ we define `deriv_n f n` to be the $n$-th derivative of polynomial $f$. $h^{[n]}$ means $h\circ h\circ h\cdots\circ h$ $n$-times.
+-/
+noncomputable def deriv_n (f : ℤ[X]) (n : ℕ) : ℤ[X] := polynomial.derivative ^[n] f
```
Is there really a need for this? It just seems to be obscuring a relatively simple definition behind a rewrite.

vihdzp commented on src/analysis/transcendental/deriv.lean:
```diff
@@ -0,0 +1,247 @@
+/-Lemma
+For any polynomial $f$ with degree $d$, the $d+1$-th derivative is zero.
+-/
+theorem deriv_too_much (f : ℤ[X]): (deriv_n f (f.nat_degree + 1)) = 0 :=
```
```suggestion
@[simp] theorem deriv_nat_degree_succ (f : ℤ[X]): (deriv_n f (f.nat_degree + 1)) = 0 :=
```

vihdzp commented on src/analysis/transcendental/deriv.lean:
```diff
@@ -0,0 +1,247 @@
+/-Lemma
+the zeroth derivative of polynomial $f$ is $f$ itself.
+-/
+lemma zeroth_deriv (f : ℤ[X]) : deriv_n f 0 = f :=
```
A lot of the theorems in this section should probably be tagged `simp`. But again, I just don't think they're needed in the first place.

vihdzp commented on src/analysis/transcendental/deriv.lean:
```diff
@@ -0,0 +1,247 @@
+            (∑ (x : ℕ) in finset.range n,
+                (polynomial.C (n.choose (x + 1):ℤ) * (deriv_n p (n - x)) * deriv_n q (x + 1) +
+                 polynomial.C (n.choose x:ℤ) * deriv_n p (n - x) * (deriv_n q (x+1)))),
+        {
```
This isn't the mathlib brace style, please check the style guide. And the spacing throughout this whole proof is off - we use 2 spaces, not 4.

vihdzp commented on src/analysis/transcendental/deriv.lean:
```diff
@@ -0,0 +1,247 @@
+        rw finset.sum_add_distrib,
+        conv_lhs {rw finset.sum_range_succ', rw finset.sum_range_succ, simp only [zeroth_deriv, nat.choose_self, one_mul, nat.choose_zero_right, int.coe_nat_zero, nat.sub_self, polynomial.C_1, int.coe_nat_succ, nat.sub_zero, zero_add]},
+
+        transitivity
```
I think you might want to use a `calc` block instead.

vihdzp commented on src/analysis/transcendental/deriv.lean:
```diff
@@ -0,0 +1,247 @@
+For a polynomial $f$ then if $n>0$, we have $f^{(n)}=f^{(n-1)}\times f'$
+-/
+
+theorem poly_pow_deriv (f : ℤ[X]) (n : ℕ) (hn : 0 < n) : (f ^ n).derivative = (polynomial.C (n:ℤ)) * (f ^ (n-1)) * f.derivative :=
```
You might find it easier to state this as 
```suggestion
theorem poly_pow_deriv (f : ℤ[X]) (n : ℕ) : (f ^ (n + 1)).derivative = (polynomial.C (n : ℤ)) * (f ^ n) * f.derivative :=
```
Also, I'm surprised this isn't immediate by the chain rule.

alreadydone replied:
There are already both versions [polynomial.derivative_pow(_succ)](https://leanprover-community.github.io/mathlib_docs/data/polynomial/derivative.html#polynomial.derivative_pow_succ) and it seems this lemma has been deleted.

vihdzp commented on src/analysis/transcendental/deriv.lean:
```diff
@@ -0,0 +1,247 @@
+  (deriv_n (polynomial.C c * f) n) = (polynomial.C c) * (deriv_n f n) :=
+begin
+  rw [deriv_n, deriv_n, polynomial.iterate_derivative_C_mul],
+end
```
```suggestion
by rw [deriv_n, deriv_n, polynomial.iterate_derivative_C_mul]
```

vihdzp commented on src/analysis/transcendental/deriv.lean:
```diff
@@ -0,0 +1,247 @@
+  rw [deriv_n, deriv_n, polynomial.iterate_derivative_C_mul],
+end
+
+theorem dvd_poly_pow_deriv (f : ℤ[X]) (n m : ℕ) (c : ℤ) (hn : 0 < n) (hm : 0 < m) :
```
```suggestion
theorem dvd_poly_pow_deriv (f : ℤ[X]) {n m : ℕ} (c : ℤ) (hn : 0 < n) (hm : 0 < m) :
```
If you have a hypothesis on some value, you should almost always make said value an implicit argument. This applies throughout.

vihdzp commented on src/analysis/transcendental/deriv.lean:
```diff
@@ -0,0 +1,247 @@
+  rw deriv_X_pow _ _ hk, ext, rw [polynomial.coeff_smul, polynomial.coeff_C_mul, smul_eq_mul],
+end
+
+lemma deriv_X_pow_too_much (n : ℕ) (k : ℕ) (hk : n < k) :
```
```suggestion
lemma deriv_X_pow_of_lt {n k : ℕ} (hk : n < k) :
```

vihdzp commented on src/analysis/transcendental/deriv.lean:
```diff
@@ -0,0 +1,247 @@
+  },
+end
+
+lemma deriv_X_sub_pow_too_much (n k : ℕ) (c : ℤ) (hk : n < k) :
```
```suggestion
lemma deriv_X_sub_pow_of_lt {n k : ℕ} (c : ℤ) (hk : n < k) :
```

vihdzp commented on src/analysis/transcendental/deriv.lean:
```diff
@@ -0,0 +1,247 @@
+    cases m, linarith,
+
+    rw [deriv_n, function.iterate_succ], simp only [function.comp_app], rw [<-deriv_n], rw poly_pow_deriv,
+    have triv : polynomial.C ↑n * f ^ (n - 1) * f.derivative = polynomial.C ↑n * (f ^ (n - 1) * f.derivative) := by ring,
```
`triv` doesn't seem like a good name for a hypothesis, first because it conflicts with the `triv` tactic, and second because these results are clearly nontrivial. Names for hypotheses starting with `h` and at most two or three characters long are preferred.

vihdzp commented on src/analysis/transcendental/deriv.lean:
```diff
@@ -0,0 +1,247 @@
+    replace hk : n < k ∨ n = k, exact lt_or_eq_of_le (nat.lt_succ_iff.mp hk),
+    cases hk,
+      have triv := IH hk,
+      rw <-deriv_succ, rw triv, simp only [polynomial.derivative_zero],
```
```suggestion
      rw ←deriv_succ, rw triv, simp only [polynomial.derivative_zero],
```
and throughout.

vihdzp reviewed (changes requested):
This is just a high-level review of the `analysis/transcendental/deriv.lean` file. This could be moved to its own standalone PR. It should also be moved elsewhere, since none of those theorems have anything to do with transcendental numbers. I didn't review the proofs themselves nor attempted to golf them.

I also suggest you brand that file as proving results on the iterated derivative. I don't know if there's other files like that already (though I'd presume so).

Ruben-VandeVelde committed:
wip

Ruben-VandeVelde committed:
wip

alexjbest commented on src/analysis/transcendental/small_lemmas.lean:
```diff
@@ -0,0 +1,75 @@
+/--
+The trivial embedding of ℤ into ℝ
+-/
+def ℤembℝ : ℤ →+* ℝ := algebra_map ℤ ℝ
```
I think we can just use `int.cast_ring_hom` throughout instead of this def

alexjbest commented on src/analysis/transcendental/small_lemmas.lean:
```diff
@@ -0,0 +1,75 @@
+/--
+f_eval_on_ℝ p is to evaluate p as a real polynomial
+-/
+def f_eval_on_ℝ (f : polynomial ℤ) (α : ℝ) : ℝ := (f.map ℤembℝ).eval α
```
This can be swapped for `polynomial.aeval`

Ruben-VandeVelde committed:
wip

Ruben-VandeVelde committed:
wip

Ruben-VandeVelde committed:
wip

Ruben-VandeVelde committed:
wip

Ruben-VandeVelde committed:
wip

Ruben-VandeVelde committed:
wip

Ruben-VandeVelde committed:
Remove deriv_n lemmas.

Ruben-VandeVelde committed:
chore(data/polynomial/derivative): merge iterated_deriv.lean into derivative.lean

iterated_deriv.lean was not used anywhere, and derivative.lean already had independent variants of several of the lemmas there, without the `iterated_deriv` indirecion. It seems better to consolidate these results.

Ruben-VandeVelde committed:
feat(data/polynomial/derivative): add more lemmas

Co-authored-by: Jujian Zhang <jujian.zhang1998@outlook.com>

Ruben-VandeVelde committed:
feat(data/polynomial/derivative): add more lemmas

Co-authored-by: Jujian Zhang <jujian.zhang1998@outlook.com>

Ruben-VandeVelde committed:
Merge branch 'polynomial-deriv-more' into some_transcendental_things-2

Ruben-VandeVelde committed:
fix

Ruben-VandeVelde committed:
wip

Ruben-VandeVelde committed:
fixup! feat(data/polynomial/derivative): add more lemmas

Ruben-VandeVelde committed:
Merge branch 'polynomial-deriv-more' into some_transcendental_things-2

Ruben-VandeVelde committed:
fixes

Ruben-VandeVelde committed:
wip

alreadydone commented on src/analysis/transcendental/e_transcendental.lean:
```diff
@@ -0,0 +1,1203 @@
+open small_lemmas e_transcendental_lemmas
+open_locale classical
+open_locale big_operators
+notation α`[X]` := polynomial α
```
Just `open_locale polynomial`.

Ruben-VandeVelde committed:
pow_transcendental

Ruben-VandeVelde committed:
Simplify pow_transcendental.

Ruben-VandeVelde committed:
generalize

Ruben-VandeVelde committed:
move

Ruben-VandeVelde committed:
tidy

Ruben-VandeVelde committed:
tidy

Ruben-VandeVelde committed:
simplify

Ruben-VandeVelde committed:
Simplify deg_f_p

Ruben-VandeVelde committed:
Simplify.

Ruben-VandeVelde committed:
wip

Ruben-VandeVelde committed:
wip

Ruben-VandeVelde committed:
Golf transform_eq

Ruben-VandeVelde committed:
fact_grows_fast'

Ruben-VandeVelde committed:
Merge remote-tracking branch 'origin/master' into some_transcendental_things-2

Ruben-VandeVelde committed:
braces

Ruben-VandeVelde committed:
wip

Ruben-VandeVelde committed:
golf

Ruben-VandeVelde committed:
fix

Ruben-VandeVelde committed:
wip

Ruben-VandeVelde committed:
move

Ruben-VandeVelde committed:
remove

Ruben-VandeVelde committed:
tidy

Ruben-VandeVelde committed:
tidy

Ruben-VandeVelde committed:
Drop deriv_n

Ruben-VandeVelde committed:
Start removing f_eval_on_ℝ

Ruben-VandeVelde committed:
Merge branch 'polynomial-deriv-more' into some_transcendental_things-2

Ruben-VandeVelde committed:
Remove f_eval_on_ℝ

Ruben-VandeVelde committed:
Drop integral_le_max_times_length

Ruben-VandeVelde committed:
Drop p_le.

Ruben-VandeVelde committed:
Drop f_p_n_succ

Ruben-VandeVelde committed:
work

Ruben-VandeVelde committed:
work

Ruben-VandeVelde committed:
work

Ruben-VandeVelde committed:
wip

Ruben-VandeVelde committed:
Use nat.desc_factorial.

Ruben-VandeVelde committed:
Merge branch 'polynomial-deriv-more' into some_transcendental_things-2

Ruben-VandeVelde committed:
update

Ruben-VandeVelde committed:
Drop deriv_exp_t_x'

Ruben-VandeVelde committed:
Golf.

Ruben-VandeVelde committed:
Golf.

Ruben-VandeVelde committed:
Move more lemmas about f_bar.

Ruben-VandeVelde committed:
Golf.

Ruben-VandeVelde committed:
Merge branch 'master' into some_transcendental_things-2

Ruben-VandeVelde committed:
Fix

Ruben-VandeVelde committed:
Golf.

Ruben-VandeVelde committed:
Golf.

Ruben-VandeVelde committed:
import

Ruben-VandeVelde committed:
open polynomial

Ruben-VandeVelde committed:
Better proof

Ruben-VandeVelde committed:
Whitespace.

Ruben-VandeVelde committed:
Whitespace.

Ruben-VandeVelde committed:
Tidy.

Ruben-VandeVelde committed:
lint

Ruben-VandeVelde committed:
Tidy.

Ruben-VandeVelde committed:
Tidy.

Ruben-VandeVelde committed:
f_bar_ineq

Ruben-VandeVelde committed:
f_bar_ineq

Ruben-VandeVelde committed:
reorganize
